### ISSUE TITLE
Naming convention changes to help deal with multiple branch commits

### ISSUE THREAD

**Issue by nycjay** [flyway/flyway#675]
I have a feature request for a change to the naming convention of sql files, to help working with commits from multiple branches.

In our shop, we have developers working on multiple branches, each with their own dev environments. The branch work could go on for a long time, and merged into trunk when ready to join the next release. Currently, each team is starting their script numbering for a given release from version 1, and going up. So, as the feature branches merge into trunk, we get conflicts on version number, and have to make manual corrections. 

Our scripts from the separate teams are very rarely related, and can be run in any order (preserving order within each feature). What would help a lot, is an enhancement to the naming convention for script files.If we could include an additional optional "prefix", we could put feature name/number in there. That would eliminate the conflicts between branches, and as long as we don't care which feature's scripts run first, we are all set. 

We did not want to use the existing prefix pattern, since we would then have to tell flyway about each feature, rather than just pointing it at a path.

**Comment by philippeehlert** [comment]
@axelfontaine what do you mean by "use one Flyway instance and metadata table per feature(-group)"? 
Could you give me an example?

**Comment by nycjay** [comment]
@philippeehlert 
We also have multiple features being developed in parallel, and ran into the same issue. Our solution was to include the Jira number in the version. 

So, if the two JIra epics are JIRA-123 and JIRA-124, the migrations scripts would be:

V123_1__JIRA-123-script1.sql
V123_2__JIRA-123-script2.sql
V123_3__JIRA-123-script3.sql
and
V124_1_JIRA-124-script1.sql
V124_2_JIRA-124-script2.sql
V124_3_JIRA-124-script3.sql

If multiple features happen to have a dependency on another (which is a rare case for us), we require that you put your script in the namespace for the JIRA ticket you are dependent on.

And, you need to enable outOfOrder mode.

**Comment by philippeehlert** [comment]
@nycjay we've been doing this for a while and we're using a timestamp as the version number, it's something like this: `V2016.10.13.09.55__script.sql`

but our main problem right now is that:
lets suppose I go from branch `master` to branch `new_feature` and create a migration there, when I'm finished, this branch goes to the code review queue, and while it is standing there to be reviewed I go back to the master branch and start coding in branch `new_feature_2` which gives me a `Detected applied migration missing on classpath` and I have to manually delete the migration from `new_feature` from `schema_version`. 

I believe that using multiple instances of Flyway would fix this problem

**Comment by nycjay** [comment]
@philippeehlert Yeah that makes sense. Our dev workflow is a bit different. When you start working on a new feature, you spin up (in AWS) a new environment, which comes w/ a clean DB (and other services needed to start up the application). You make your branch off of master, which comes with all migration scripts you need to get your environment up to date with master (via Flyway). When you are done developing a feature at it is tested / merged into master, then you destroy that environment, and spin up an new one for your next feature. That makes things a whole lot easier...

But yeah, multiple Flyway tables would also fix the problem, you just end up with a bunch of noise/orphaned tables in your DB.
